src/trie/merkle_tree.rs

```diff
@@ -1312,6 +1334,463 @@ pub(crate) fn bytes_to_bitvec(bytes: &[u8]) -> BitVec<u8, Msb0> {
+                let ktrunc = &truncate(k);
+                let kfelt0 = Felt::from_bytes_be_slice(k);
+                let kfelt1 = Felt::from_bytes_be_slice(ktrunc.as_raw_slice());
+
+                // quick sanity check to make sure truncating a key does not remove any data
+                assert_eq!(kfelt0, kfelt1);
+
+                let v = &Felt::from_bytes_be_slice(v);
+                assert!(bonsai.insert(contract_address, ktrunc, v).is_ok());
+            }
+        }
+        assert!(bonsai.commit(BasicId::new(0)).is_ok());
+
+        // aggreates all storage changes to their latest state
+        // (replacements are takent into account)
+        let mut storage_map = IndexMap::<Vec<u8>, IndexMap<Felt, Felt>>::new();
+        for (contract_address, storage) in blocks.clone() {
+            let map = storage_map
+                .entry(contract_address.to_vec())
+                .or_insert(IndexMap::new());
+
+            for (k, v) in storage {
+                let k = Felt::from_bytes_be_slice(k);
+                let v = Felt::from_bytes_be_slice(v);
+                map.insert(k, v);
+            }
+        }
+
+        // checks for each contract if the original key can be reconstructed
+        // from the data stored in the db
+        for (contract_address, storage) in storage_map.iter() {
+            log::info!(
+                "contract address (read): {:#064x}",
+                Felt::from_bytes_be_slice(contract_address)
+            );
+
+            let keys = bonsai.get_keys(contract_address).unwrap();
+            log::debug!("{keys:?}");
+            for k in keys {
+                // if all has gone well, the db should contain the first 251 bits of the key,
+                // which should represent the entirety of the data
+                let k = Felt::from_bytes_be_slice(&k);
+                log::info!("looking for key: {k:#064x}");
+
+                assert!(storage.contains_key(&k));
+            }
+        }
+
+        // makes sure retrieving key-value pairs works for each contract
+        for (contract_address, storage) in storage_map.iter() {
+            log::info!(
+                "contract address (read): {:#064x}",
+                Felt::from_bytes_be_slice(contract_address)
+            );
+
+            let kv = bonsai.get_key_value_pairs(contract_address).unwrap();
+            log::debug!("{kv:?}");
+            for (k, v) in kv {
+                let k = Felt::from_bytes_be_slice(&k);
+                let v = Felt::from_bytes_be_slice(&v);
+                log::info!("checking for key-value pair:({k:#064x}, {v:#064x})");
+
+                assert_eq!(*storage.get(&k).unwrap(), v);
+            }
+        }
+    }
+
+    fn str_to_felt_bytes(hex: &str) -> [u8; 32] {
+        Felt::from_hex(hex).unwrap().to_bytes_be()
+    }
+
+    fn truncate(key: &[u8]) -> BitVec<u8, Msb0> {
+        key.view_bits()[5..].to_owned()
+    }
     // use crate::{
```

Comment:
Can we do a bit of cleanup here? Or do we keep it for later?

Reply:
This is an artifact from a previous version of the code where there were multiple calls to this function. I think it can be a good idea to keep it in though for future tests as skipping the first 5 bits of a key can be quite error prone when done manually and repeatedly.